server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java

```diff
@@ -0,0 +1,247 @@
+        public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+            joinCallback.onSuccess();
+        }
+
+        @Override
+        public String toString() {
+            return "JoinTaskListener{task=" + task + "}";
+        }
+    }
+
+
+    private interface JoinAccumulator {
+        void handleJoinRequest(DiscoveryNode sender, JoinCallback joinCallback);
+
+        default void failPendingJoins(String reason) {
+        }
+
+        default void submitPendingJoins() {
+        }
+    }
+
+    private class LeaderJoinAccumulator implements JoinAccumulator {
+        @Override
+        public void handleJoinRequest(DiscoveryNode sender, JoinCallback joinCallback) {
+            final JoinTaskExecutor.Task task = new JoinTaskExecutor.Task(sender, "join existing leader");
+            masterService.submitStateUpdateTask("node-join", task, ClusterStateTaskConfig.build(Priority.URGENT),
+                joinTaskExecutor, new JoinTaskListener(task, joinCallback));
+        }
+
+        @Override
+        public String toString() {
+            return "LeaderJoinAccumulator";
+        }
+    }
+
+    private static class FollowerJoinAccumulator implements JoinAccumulator {
+        @Override
+        public void handleJoinRequest(DiscoveryNode sender, JoinCallback joinCallback) {
+            joinCallback.onFailure(new CoordinationStateRejectedException("join target is a follower"));
+        }
+
+        @Override
+        public String toString() {
+            return "FollowerJoinAccumulator";
+        }
+    }
+
+    private class CandidateJoinAccumulator implements JoinAccumulator {
+
+        private final Map<DiscoveryNode, JoinCallback> joinRequestAccumulator = new HashMap<>();
+
+        @Override
+        public void handleJoinRequest(DiscoveryNode sender, JoinCallback joinCallback) {
+            JoinCallback prev = joinRequestAccumulator.put(sender, joinCallback);
+            if (prev != null) {
+                prev.onFailure(new CoordinationStateRejectedException("received a newer join from " + sender));
+            }
+        }
+
+        @Override
+        public void failPendingJoins(String reason) {
+            joinRequestAccumulator.values().forEach(joinCallback -> joinCallback.onFailure(new CoordinationStateRejectedException(reason)));
+        }
+
+        @Override
+        public void submitPendingJoins() {
+            final Map<JoinTaskExecutor.Task, ClusterStateTaskListener> pendingAsTasks = new HashMap<>();
+            joinRequestAccumulator.forEach((key, value) -> {
+                final JoinTaskExecutor.Task task = new JoinTaskExecutor.Task(key, "elect leader");
+                pendingAsTasks.put(task, new JoinTaskListener(task, value));
+            });
+
+            final String stateUpdateSource = "elected-as-master ([" + pendingAsTasks.size() + "] nodes joined)";
+
+            pendingAsTasks.put(JoinTaskExecutor.BECOME_MASTER_TASK, (source, e) -> {
```

Comment:
Should really ask for `toString()`s on these handlers too, although this adds noise.

Reply:
The handler is not logged anywhere, so adding toString adds little.